File: backend/routers/alertas.py
from fastapi import APIRouter, HTTPException, Depends
from backend.core.database import get_db
from backend.schemas.alertas_schema import AlertasCreate, AlertasUpdate, AlertasSchema
from typing import List
from bson import ObjectId

router = APIRouter(prefix="/api/alertas", tags=["alertas"])

async def get_db_dependency():
    return get_db()

def _normalize_priority(value):
    raw = str(value or "").strip().lower()
    mapping = {
        "crítico": "critica",
        "critico": "critica",
        "critical": "critica",
        "alta": "alta",
        "high": "alta",
        "media": "media",
        "média": "media",
        "medium": "media",
        "baixa": "baixa",
        "low": "baixa",
    }
    return mapping.get(raw, raw or "media")


def _normalize_status(value):
    raw = str(value or "").strip().lower()
    mapping = {
        "lido": "lido",
        "read": "lido",
        "resolvido": "resolvido",
        "resolved": "resolvido",
        "arquivado": "arquivado",
        "archived": "arquivado",
        "pendente": "pendente",
        "pending": "pendente",
    }
    return mapping.get(raw, raw or "pendente")


def _normalize_alert(item):
    payload = item.get("data") if isinstance(item.get("data"), dict) else {}
    prioridade = _normalize_priority(item.get("prioridade") or payload.get("prioridade") or item.get("nivel") or payload.get("nivel"))
    status = _normalize_status(item.get("status") or payload.get("status") or ("resolvido" if item.get("resolvido") or payload.get("resolvido") else "pendente"))
    resolved = bool(item.get("resolvido") or payload.get("resolvido") or status in {"resolvido", "arquivado"})
    read = bool(item.get("lido") or payload.get("lido") or status == "lido")
    created_at = item.get("created_at") or payload.get("created_at") or item.get("createdAt") or payload.get("createdAt") or item.get("timestamp") or payload.get("timestamp")

    return {
        **item,
        "id": str(item.get("_id") or item.get("id") or ""),
        "titulo": item.get("titulo") or payload.get("titulo") or item.get("mensagem") or payload.get("mensagem") or "Alerta",
        "descricao": item.get("descricao") or payload.get("descricao") or item.get("mensagem") or payload.get("mensagem") or "",
        "prioridade": prioridade,
        "status": status,
        "lido": read,
        "resolvido": resolved,
        "data": created_at,
        "payload": payload,
    }


@router.get("/")
async def list_items(db=Depends(get_db_dependency)):
    items = await db["alertas"].find().to_list(100)
    return [_normalize_alert(item) for item in items]

@router.post("/", response_model=AlertasSchema)
async def create_item(item: AlertasCreate, db=Depends(get_db_dependency)):
    result = await db["alertas"].insert_one(item.dict())
    item_dict = item.dict()
    item_dict["id"] = str(result.inserted_id)
    item_dict["ativo"] = True
    return item_dict

File: frontend/src/lib/alertas.js
const normalizeText = (value, fallback = "") => {
  if (typeof value !== "string") return fallback;
  const trimmed = value.trim();
  return trimmed || fallback;
};

const normalizePriority = (value) => {
  const raw = normalizeText(value).toLowerCase();
  const mapping = {
    crítico: "critica",
    critico: "critica",
    critical: "critica",
    alta: "alta",
    high: "alta",
    media: "media",
    média: "media",
    medium: "media",
    baixa: "baixa",
    low: "baixa",
  };

  return mapping[raw] || raw || "media";
};

const normalizeStatus = (value) => {
  const raw = normalizeText(value).toLowerCase();
  const mapping = {
    lido: "lido",
    read: "lido",
    resolvido: "resolvido",
    resolved: "resolvido",
    arquivado: "arquivado",
    archived: "arquivado",
    pendente: "pendente",
    pending: "pendente",
  };

  return mapping[raw] || raw || "pendente";
};

const isResolvedAlert = (alerta) => {
  const status = normalizeStatus(alerta?.status || alerta?.data?.status);
  return Boolean(alerta?.resolvido || alerta?.data?.resolvido || status === "resolvido" || status === "arquivado");
};

const isReadAlert = (alerta) => {
  const status = normalizeStatus(alerta?.status || alerta?.data?.status);
  return Boolean(alerta?.lido || alerta?.data?.lido || status === "lido");
};

const isUnreadAlert = (alerta) => !isReadAlert(alerta) && !isResolvedAlert(alerta);

const isCriticalAlert = (alerta) => {
  const prioridade = normalizePriority(alerta?.prioridade || alerta?.data?.prioridade || alerta?.nivel || alerta?.data?.nivel);
  return ["critica", "alta"].includes(prioridade) && !isResolvedAlert(alerta);
};

const normalizeAlerta = (item = {}) => {
  const nested = typeof item?.data === "object" && item.data ? item.data : {};
  const prioridade = normalizePriority(item.prioridade || nested.prioridade || item.nivel || nested.nivel);
  const status = normalizeStatus(item.status || nested.status || (item.resolvido || nested.resolvido ? "resolvido" : item.lido || nested.lido ? "lido" : "pendente"));
  const resolved = isResolvedAlert({ ...item, prioridade, status });
  const read = isReadAlert({ ...item, prioridade, status });

  return {
    ...item,
    id: item.id || item._id || nested.id || "",
    tipo: normalizeText(item.tipo || nested.tipo, "sistema"),
    prioridade,
    titulo: normalizeText(item.titulo || nested.titulo || item.mensagem || nested.mensagem, "Alerta"),
    descricao: normalizeText(item.descricao || nested.descricao || item.mensagem || nested.mensagem, ""),
    empresa: item.empresa || nested.empresa || null,
    tempo: normalizeText(item.tempo || nested.tempo, ""),
    status,
    lido: read,
    resolvido: resolved,
    data: item.data || item.created_at || item.createdAt || item.timestamp || nested.data || nested.created_at || nested.createdAt || null,
  };
};

const normalizeAlertas = (items) => {
  if (!Array.isArray(items)) return [];
  return items.map(normalizeAlerta);
};

const countBadgeAlertas = (items) => normalizeAlertas(items).filter(isUnreadAlert).length;

const countCriticalAlertas = (items) => normalizeAlertas(items).filter(isCriticalAlert).length;

const formatBadgeCount = (count) => {
  const total = Number(count);
  if (!Number.isFinite(total) || total <= 0) return 0;
  return total > 99 ? "99+" : total;
};

export {
  countBadgeAlertas,
  countCriticalAlertas,
  formatBadgeCount,
  isCriticalAlert,
  isReadAlert,
  isResolvedAlert,
  isUnreadAlert,
  normalizeAlerta,
  normalizeAlertas,
  normalizePriority,
  normalizeStatus,
};
